### PR TITLE
Moved EGL into a separate context

### PR DESCRIPTION
**Pull request by davidpiuva**
First step of separating EGL from OpenGL.

### CODE REVIEW

**Comment by davidpiuva** [comment]
@emilwestergren Peer review?

File: source/draw/gpu/opengl/backend/gles3/Gles3Context.ooc [hunk 1]
```diff
@@ -9,126 +9,20 @@
...
 import external/gles3
 import Gles3Debug, Gles3Fence, Gles3FramebufferObject, Gles3Quad, Gles3Renderer, Gles3ShaderProgram, Gles3Texture, Gles3VolumeTexture, Gles3VertexArrayObject, Gles3IndexBufferObject
 
 Gles3Context: class extends GLContext {
-	_eglContext: Pointer
-	_eglSurface: Pointer
-	_contextCount := static 0
-	_mutex := static Mutex new()
+	displayContext := EglDisplayContext new()
+	getDisplayContextSafely: func -> EglDisplayContext { this == null ? null : this displayContext }
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/backend/gles3/Gles3Context.ooc]: A bit ugly to compare this to null but the alternatives are to duplicate the code or make the getter static.

File: source/draw/gpu/opengl/backend/EglDisplayContext.ooc [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+		}
+		This validate(eglMakeCurrent(this _eglDisplay, this _eglSurface, this _eglSurface, this _eglContext), EGL_TRUE, "eglMakeCurrent")
+		true
+	}
+	_initializeDisplay: func (display: Pointer) {
+		this _eglDisplay = eglGetDisplay(display)
+		This validate(this _eglDisplay != EGL_NO_DISPLAY, "eglGetDisplay")
+		This validate(eglInitialize(this _eglDisplay, null, null), EGL_TRUE, "eglInitialize")
+	}
+	_generate: func (display: Pointer, nativeBackend: Long, sharedContext: This) -> Bool {
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/backend/EglDisplayContext.ooc]: I do not know if these should be private or public when only called from the OpenGL context creation.

Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/EglDisplayContext.ooc]: If you want we can replace the "return null on failure" philosophy with good error checking that throws exception instead (we don't handle the null return anyway) Then you don't need these ugly `_generate` functions, you can simply overload constructors.

File: source/draw/gpu/opengl/backend/gles3/Gles3Context.ooc (hunk 1 above)
Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/gles3/Gles3Context.ooc]: Ok, as a temporary solution I can approve it.

File: source/draw/gpu/opengl/backend/gles3/Gles3Context.ooc [hunk 3]
```diff
@@ -9,126 +9,20 @@
...
+import ../[GLContext, GLTexture, GLVertexArrayObject, GLIndexBufferObject, EglDisplayContext]
 import external/gles3
 import Gles3Debug, Gles3Fence, Gles3FramebufferObject, Gles3Quad, Gles3Renderer, Gles3ShaderProgram, Gles3Texture, Gles3VolumeTexture, Gles3VertexArrayObject, Gles3IndexBufferObject
 
 Gles3Context: class extends GLContext {
-	_eglContext: Pointer
-	_eglSurface: Pointer
-	_contextCount := static 0
-	_mutex := static Mutex new()
+	displayContext := EglDisplayContext new()
```
Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/gles3/Gles3Context.ooc]: Public member variable? Should be private or have a getter property.

File: source/draw/gpu/opengl/backend/EglDisplayContext.ooc (hunk 2 above)
Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/EglDisplayContext.ooc]: Keep shared context feature, try to create a some nice constructor overloads. Either by defaulting shared context to null or overload constructors and demand that the user makes the choice of the constructor.

Comment (davidpiuva) [inline source/draw/gpu/opengl/backend/EglDisplayContext.ooc]: I tried overloads for `Gles3Context create` but in order to use them, I had to duplicate `GLContext createContext` to have any use of the new overloads. Default values for internal things like EGL might not be needed since it should be hidden inside the OpenGL context.

Comment (davidpiuva) [inline source/draw/gpu/opengl/backend/EglDisplayContext.ooc]: I might try just having constructors for Gles3Context then since _generate is a mess.

Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/EglDisplayContext.ooc]: `Gles3Context create` can be replaced with constructor as well.

Comment (davidpiuva) [inline source/draw/gpu/opengl/backend/EglDisplayContext.ooc]: That would be cleaner.

File: source/draw/gpu/opengl/backend/gles3/Gles3Context.ooc [hunk 4]
```diff
@@ -191,25 +85,13 @@ Gles3Context: class extends GLContext {
...
 		version(debugGL) { Debug print("Creating OpenGL Context") }
-		result := This new()
-		result _generate(display, nativeBackend, sharedContext) ? result : null
+		result := This new(EglDisplayContext new(display, nativeBackend, sharedContext getDisplayContextSafely()))
 	}
 	create: static func ~pbufferShared (sharedContext: This = null) -> This {
 		version(debugGL) { Debug print("Creating OpenGL Context") }
-		result := This new()
-		result _generate(sharedContext) ? result : null
+		result := This new(EglDisplayContext new(sharedContext getDisplayContextSafely()))
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/backend/gles3/Gles3Context.ooc]: This looks cleaner but I do not know if this will work when GLES3 has to fall back to GLES2 since the constructor cannot have a boolean return value.

Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/gles3/Gles3Context.ooc]: So keep the `create` for Gles3Context

Comment (davidpiuva) [inline source/draw/gpu/opengl/backend/gles3/Gles3Context.ooc]: Okay

File: source/draw/gpu/opengl/backend/gles3/Gles3Context.ooc [hunk 5]
```diff
@@ -9,126 +9,20 @@
 version(!gpuOff) {
 use base
 use geometry
-import ../egl/egl
-import ../[GLContext, GLTexture, GLVertexArrayObject, GLIndexBufferObject]
+import ../[GLContext, GLTexture, GLVertexArrayObject, GLIndexBufferObject, EglDisplayContext]
 import external/gles3
 import Gles3Debug, Gles3Fence, Gles3FramebufferObject, Gles3Quad, Gles3Renderer, Gles3ShaderProgram, Gles3Texture, Gles3VolumeTexture, Gles3VertexArrayObject, Gles3IndexBufferObject
```
Comment (marcusnaslund) [inline source/draw/gpu/opengl/backend/gles3/Gles3Context.ooc]: Huh, I guess that's not true always

File: source/draw/gpu/opengl/backend/EglDisplayContext.ooc [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+ * Copyright (C) 2016 magic-lang
+ *
+ * This software may be modified and distributed under the terms
+ * of the MIT license.  See the LICENSE file for details.
+ */
+
+version(!gpuOff) {
+use base
+use geometry
+import egl/egl
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/backend/EglDisplayContext.ooc]: It was like that from where I copied it from so any change in behaviour should be a separate PR.

Comment (marcusnaslund) [inline source/draw/gpu/opengl/backend/EglDisplayContext.ooc]: yeah and as my next comment shows, maybe "generally outside in other files" isn't even true
